src/flush.rs

```diff
@@ -114,27 +120,55 @@ impl DbInner {
                             WalFlushThreadMsg::Shutdown => {
                                 // Stop the thread.
                                 _ = this.flush().await;
-                                return
+                                break Ok(())
                             },
                             WalFlushThreadMsg::FlushImmutableWals => {
                                 let result = this.flush().await;
                                 if let Err(err) = &result {
                                     error!("error from wal flush: {err}");
-                                    this.set_error_if_none(err.clone());
+                                    break Err(err.clone());
                                 }
 
                                 if let Some(rsp_sender) = rsp_sender {
                                     let res = rsp_sender.send(result);
                                     if let Err(Err(err)) = res {
                                         error!("error sending flush response: {err}");
-                                        this.set_error_if_none(err);
                                     }
                                 }
                             },
                         }
                     }
                 }
+            };
+
+            let pending_result = result.clone().and_then(|_| Err(BackgroundTaskShutdown));
+            while !rx.is_empty() {
+                let (rsp_sender, _) = rx.recv().await.expect("channel unexpectedly closed");
+                if let Some(rsp_sender) = rsp_sender {
+                    let _ = rsp_sender.send(pending_result.clone());
+                }
             }
-        }))
+
+            info!("wal flush thread exiting with {:?}", result);
+            result
+        };
+
+        let this = Arc::clone(self);
+        Some(spawn_bg_task(
+            tokio_handle,
+            move |err| {
+                info!("flush task exited with {:?}", err);
+                // notify any waiters about the failure
+                let mut state = this.state.write();
+                state.record_fatal_error(err.clone());
```

Comment:
> Just to confirm: Since we're using WatchableOnceCell in KVTable and ImmutableMemtable below, the notify_durable calls here will not notify of an error if the listeners have already been notified of a success, right?

Yes, this is intentional. Once we've recorded that a memtable is flushed/durable it doesn't make sense to record an error.

> this code seems somewhat brittle to refactors where we notify_durable in flush_imm_wals and then later notify of a failure of some other call fails afterward.

This is actually how it works even in the current impl. When the flush or memtable task fails, the cleanup_fn will set the error in all wal/memtables by calling notify_durable/notify_flush with an error. This relies on notify_durable/notify_flush not overwriting a result if it's already set.